src/config/base.cpp

```diff
@@ -651,53 +651,65 @@ LIBSESSION_EXPORT config_string_list* config_merge(
         const unsigned char** configs,
         const size_t* lengths,
         size_t count) {
-    auto& config = *unbox(conf);
-    std::vector<std::pair<std::string, ustring_view>> confs;
-    confs.reserve(count);
-    for (size_t i = 0; i < count; i++)
-        confs.emplace_back(msg_hashes[i], ustring_view{configs[i], lengths[i]});
-
-    return make_string_list(config.merge(confs));
+    try {
+        auto& config = *unbox(conf);
+        std::vector<std::pair<std::string, ustring_view>> confs;
+        confs.reserve(count);
+        for (size_t i = 0; i < count; i++)
+            confs.emplace_back(msg_hashes[i], ustring_view{configs[i], lengths[i]});
+
+        return make_string_list(config.merge(confs));
+    } catch (const std::exception& e) {
+        copy_c_str(conf->_error_buf, e.what());
+        conf->last_error = conf->_error_buf;
+    }
+    return nullptr;
```

Comment:
We could slightly genericize this error handling with a pair of templated functions like this:

```C++
// No return value: accepts void and pointer returns; pointer returns will become nullptr on error
template <typename Call>
auto exc_to_error(config_object* conf, Call& f) {
    using Ret = decltype(f());
    try {
        return f();
    } catch (const std::exception& e) {
        copy_c_str(conf->_error_buf, e.what());
        conf->last_error = conf->_error_buf;
    }
    if constexpr (std::is_pointer_v<Ret>)
        return nullptr;
    else static_assert(std::is_void_v<Ret>, "Don't know how to return an error value!");
}

// Accepts callbacks with value returns on errors: returns `f()` on success, `error_return` on exception
template <typename Call, typename Ret>
Ret exc_to_error(config_object* conf, Call& f, Ret error_return) {
    try {
        return f();
    } catch (const std::exception& e) {
        copy_c_str(conf->_error_buf, e.what());
        conf->last_error = conf->_error_buf;
    }
    return error_return;
}
```

and then can replace all of these `try-stuff-catch-set-error` blocks with:
```C++
LIBSESSION_EXPORT config_string_list* config_merge(
        config_object* conf,
        const char** msg_hashes,
        const unsigned char** configs,
        const size_t* lengths,
        size_t count) {
    return exc_to_error(conf, [&] {
        auto& config = *unbox(conf);
        std::vector<std::pair<std::string, ustring_view>> confs;
        confs.reserve(count);
        for (size_t i = 0; i < count; i++)
            confs.emplace_back(msg_hashes[i], ustring_view{configs[i], lengths[i]});

        return make_string_list(config.merge(confs));
    });
```

It's not a *huge* gain, but it does reduce the repetitiveness a bit.

The first version of the function will work in most places (with `void` or pointer returns -- the latter returns nullptr on exception), but if there's some version that returns a value (say an int) then the second one lets you do:
```C++
return exc_to_error(conf, [&] { return get_some_integer_that_might_throw(); }, -1);
```
to specify the value to return if an exception happens.

Reply:
`exc_to_error` is probably not a great name.  Maybe `wrap_exceptions` or something?

Reply:
No worries, I've added these in and updated all of the C functions which call throwing C++ functions to use it